craft_parts/sequencer.py

```diff
@@ -81,3 +84,103 @@ def _add_step_actions(
     ) -> None:
         # TODO: implement this stub
         pass
+
+    def _process_dependencies(self, part: Part, step: Step) -> None:
+        prerequisite_step = steps.dependency_prerequisite_step(step)
+        if not prerequisite_step:
+            return
+
+        all_deps = parts.part_dependencies(part.name, part_list=self._part_list)
+
+        deps = {p for p in all_deps if self._sm.should_step_run(p, prerequisite_step)}
+        for dep in deps:
+            self._add_all_actions(
+                target_step=prerequisite_step,
+                part_names=[dep.name],
+                reason=f"required to {_step_verb[step]} {part.name!r}",
+            )
+
+    def _run_step(
+        self,
+        part: Part,
+        step: Step,
+        *,
+        reason: Optional[str] = None,
+        rerun: bool = False,
+    ) -> None:
+        self._process_dependencies(part, step)
+
+        if rerun:
+            self._add_action(part, step, action_type=ActionType.RERUN, reason=reason)
+        else:
+            self._add_action(part, step, reason=reason)
+
+        state: states.StepState
+        part_properties = part.spec.marshal()
+
+        if step == Step.PULL:
+            state = states.PullState(
+                part_properties=part_properties,
+                project_options=self._project_info.project_options,
+                assets={},  # TODO: obtain pull assets
+            )
+
+        elif step == Step.BUILD:
+            state = states.BuildState(
+                part_properties=part_properties,
+                project_options=self._project_info.project_options,
+                assets={},  # TODO: obtain build assets
+            )
+
+        elif step == Step.STAGE:
+            state = states.StageState(
+                part_properties=part_properties,
+                project_options=self._project_info.project_options,
+                files=set(),
+                directories=set(),
+            )
+
+        else:
```

Comment:
should be elif, consider we may add new steps, else should raise a RuntimeError

Reply:
Raising RuntimeError in case of invalid step. Added similar handling to state loading for consistency.